File: 백준/Silver/9012. 괄호/괄호.cc
#include <bits/stdc++.h>
using namespace std;

int n;
string s;

bool check(string s) {
  stack<char> stk;
  for (char c : s) {
    if (c == '(')
      stk.push(c);
    else {
      if (!stk.empty())
        stk.pop();
      else
        return false;
    }
  }
  return stk.empty();
}

int main() {
  cin >> n;
  for (int i = 0; i < n; i++) {
    cin >> s;
    if (check(s))
      cout << "YES\n";
    else
      cout << "NO\n";
  }

  return 0;
}

// Stack은 {짝짓기, 폭발, 아름다운 괄호} 키워드 기억
// 중첩 반복문에서 i j 잘 구분해서 작성하기

File: 백준/Silver/9012. 괄호/괄호.js
const fs = require('fs');
const input = fs.readFileSync('/dev/stdin').toString().trim().split('\n');

const n = Number(input[0]);
const arr = input.slice(1);

const answer = (n, arr) => {
  arr.map((s) => {
    const stk = [];

    [...s].map((c) => {
      if (c === '(' || stk.length === 0) {
        stk.push(c);
      } else if (c === ')' && stk.length !== 0) {
        if (stk[stk.length - 1] === '(') {
          stk.pop();
        }
      }
    });

    if (stk.length !== 0) {
      console.log('NO');
    } else {
      console.log('YES');
    }
  });
};

answer(n, arr);
